File: android/app/src/main/java/com/ras/data/webrtc/SdpValidator.kt
package com.ras.data.webrtc

/**
 * Validates SDP content for required fields.
 *
 * Used to catch configuration errors early rather than mysterious timeouts.
 */
object SdpValidator {

    /**
     * Minimum number of ICE candidates expected in a valid offer/answer.
     * At least one host candidate should always be present.
     */
    const val MIN_EXPECTED_CANDIDATES = 1

    /**
     * Validate that SDP contains ICE candidates.
     *
     * @param sdp The SDP string to validate
     * @param description Human-readable description for error messages
     * @throws IllegalStateException if no candidates found
     */
    fun requireCandidates(sdp: String, description: String = "SDP") {
        val candidateCount = countCandidates(sdp)
        if (candidateCount < MIN_EXPECTED_CANDIDATES) {
            throw IllegalStateException(
                "$description contains no ICE candidates. " +
                "This usually means ICE gathering didn't complete before SDP was sent. " +
                "Candidates found: $candidateCount"
            )
        }
    }

    /**
     * Count ICE candidates in SDP.
     */
    fun countCandidates(sdp: String): Int {
        return sdp.lines().count { it.startsWith("a=candidate:") }
    }

    /**
     * Check if SDP contains at least one host candidate (local network).
     * Looks for "typ host" pattern which may be at end of line.
     */
    fun hasHostCandidate(sdp: String): Boolean {
        return sdp.lines().any {
            it.startsWith("a=candidate:") && it.contains(" host")
        }
    }

    /**
     * Check if SDP contains server-reflexive candidate (STUN).
     * Looks for "typ srflx" pattern.
     */
    fun hasServerReflexiveCandidate(sdp: String): Boolean {
        return sdp.lines().any {
            it.startsWith("a=candidate:") && it.contains(" srflx")
        }
    }

    /**
     * Check if SDP contains relay candidate (TURN).
     * Looks for "typ relay" pattern.
     */
    fun hasRelayCandidate(sdp: String): Boolean {
        return sdp.lines().any {
            it.startsWith("a=candidate:") && it.contains(" relay")
        }
    }

    /**
     * Extract all candidates from SDP for debugging.
     */
    fun extractCandidates(sdp: String): List<String> {
        return sdp.lines().filter { it.startsWith("a=candidate:") }
    }
}

File: daemon/src/ras/sdp_validator.py
"""SDP validation utilities.

Validates SDP content to catch configuration errors early.
"""

from dataclasses import dataclass


@dataclass
class SdpValidationResult:
    """Result of SDP validation."""

    is_valid: bool
    candidate_count: int
    has_host: bool
    has_srflx: bool
    has_relay: bool
    errors: list[str]


def validate_sdp(sdp: str, description: str = "SDP") -> SdpValidationResult:
    """Validate SDP contains required fields.

    Args:
        sdp: The SDP string to validate
        description: Human-readable description for error messages

    Returns:
        SdpValidationResult with validation details
    """
    errors = []

    # Count candidates by type
    candidates = [line for line in sdp.split("\n") if line.startswith("a=candidate:")]
    candidate_count = len(candidates)

    # Check for candidate types - they appear as "typ host", "typ srflx", etc.
    # Need to handle end-of-line (no trailing space)
    has_host = any(" host" in c for c in candidates)
    has_srflx = any(" srflx" in c for c in candidates)
    has_relay = any(" relay" in c for c in candidates)

    if candidate_count == 0:
        errors.append(f"{description} contains no ICE candidates")

    is_valid = len(errors) == 0

    return SdpValidationResult(
        is_valid=is_valid,
        candidate_count=candidate_count,
        has_host=has_host,
        has_srflx=has_srflx,
        has_relay=has_relay,
        errors=errors,
    )


def require_candidates(sdp: str, description: str = "SDP") -> None:
    """Validate SDP contains ICE candidates, raise if not.

    Args:
        sdp: The SDP string to validate
        description: Human-readable description for error messages

    Raises:
        ValueError: If no candidates found
    """
    result = validate_sdp(sdp, description)
    if not result.is_valid:
        raise ValueError(
            f"{description} validation failed: {', '.join(result.errors)}. "
            "This usually means ICE gathering didn't complete before SDP was sent."
        )


def extract_candidates(sdp: str) -> list[str]:
    """Extract all ICE candidates from SDP.

    Args:
        sdp: The SDP string

    Returns:
        List of candidate lines
    """
    return [line for line in sdp.split("\n") if line.startswith("a=candidate:")]
